### ISSUE TITLE
Replace Hardcoded Anomaly Detection with ML/AI Detection

### ISSUE THREAD

**Issue by AhmedIkram05** [AhmedIkram05/laad#9]
Replaces the current rule-based `anomaly_detector.py` with a full XGBoost ML pipeline.

---

## ⚠️ CRITICAL IMPLEMENTATION NOTES FOR COPILOT

Read these before writing any code. Five bugs are pre-identified that will silently produce garbage or hard errors if not addressed.

---

### Bug 1 — CRITICAL (silent): JSONB type mismatch in `feature_engineering.py`

psycopg2 **automatically deserialises PostgreSQL JSONB columns as Python `dict`**, not `str`. Any `isinstance(p, str)` check will always be `False`, silently returning fallback values (0 / False) for every payload-derived feature.

**Required fix:** Add this helper near the top of `feature_engineering.py` and route every payload feature through it — no inline `isinstance(p, str)` lambdas anywhere:

```python
import json

def parse_payload(p):
    """psycopg2 returns JSONB as dict already; handle both for test-fixture compatibility."""
    if isinstance(p, dict):
        return p
    if isinstance(p, str):
        try:
            return json.loads(p)
        except (ValueError, TypeError):
            return {}
    return {}
```

Every lambda in `extract_features` that reads from `raw_payload` must call `parse_payload(p).get("key", fallback)` — never `p.strip().startswith("{")` or `isinstance(p, str)`.

---

### Bug 2 — CRITICAL (silent): `_anomaly_tag` missing from A1–A6 injectors in `anomaly_injectors.py`

`extract_label()` keys off the `_anomaly_tag` field. If any injector omits it, that anomaly class will have **zero positive training examples** and XGBoost will never learn to detect it.

**Every injector MUST set this field on every emitted record:**

```python
record["_anomaly_tag"] = "A1"   # change to A2–A6 per injector
```

| Injector class | Required `_anomaly_tag` |
|---|---|
| Spike injector | `"A1"` |
| Dropout injector | `"A2"` |
| Latency injector | `"A3"` |
| Reorder injector | `"A4"` |
| Corrupt-payload injector | `"A5"` |
| Schema-drift injector | `"A6"` |

Normal (non-injected) records must **not** have `_anomaly_tag` set (or set to `None`/`"normal"`) so `extract_label()` returns the correct negative class.

---

### Bug 3 — CRITICAL (hard error): Wrong import path in `ml_detector.py`

```python
# ❌ WRONG — raises ModuleNotFoundError
from anomaly_detector import AnomalyDetector

# ✅ CORRECT — matches Phase 1 file structure
from backend.src.anomaly_detection.anomaly_detector import AnomalyDetector
```

---

### Bug 4 — Minor but will error: Dead `dateutil` code in `train.py`

psycopg2 deserialises `TIMESTAMPTZ` columns as **timezone-aware Python `datetime` objects** automatically. The `isinstance(r["timestamp"], str)` check and `parse_dt` call are dead code that will also fail at install time because `python-dateutil` is not listed in requirements.

**Delete both of these lines entirely:**

```python
# DELETE this import
from dateutil.parser import parse as parse_dt

# DELETE this conditional — replace the whole expression with just:
timestamp = r["timestamp"]
# (not: parse_dt(r["timestamp"]) if isinstance(r["timestamp"], str) else r["timestamp"])
```

Do **not** add `python-dateutil` to requirements.

---

### Bug 5 — Low risk: Wrong MLflow Docker image registry in `docker-compose.yml`

MLflow's official image is on **Docker Hub**, not ghcr.io:

```yaml
# ❌ WRONG
image: ghcr.io/mlflow/mlflow:v2.12.0

# ✅ CORRECT
image: mlflow/mlflow:v2.12.0
```

---

## Implementation Checklist

- [ ] `parse_payload()` helper in `feature_engineering.py`; all lambdas use it
- [ ] Every A1–A6 injector sets `record["_anomaly_tag"]`
- [ ] `ml_detector.py` uses `from backend.src.anomaly_detection.anomaly_detector import AnomalyDetector`
- [ ] `train.py` has no `dateutil` import; `timestamp = r["timestamp"]` directly
- [ ] `docker-compose.yml` MLflow service uses `image: mlflow/mlflow:v2.12.0`

**Comment by AhmedIkram05** [comment]
- Built a fully local RAG-based diagnostic assistant (LangChain, ChromaDB, Ollama) enabling natural language
querying of ATM log history - architected as an air-gapped system ensuring no log data leaves the local
network, directly aligned with banking-grade data privacy requirements
- Implemented semantic chunking via LangChain's SemanticChunker for context-aware log segmentation, with
nomic-embed-text embeddings and ChromaDB as the vector store; inference served locally by llama3.1:8b
- Evaluated retrieval and generation quality using an LLM-as-judge pipeline - scoring relevance, faithfulness,
and answer completeness across a held-out set of diagnostic queries
